### DISCUSSION TITLE
Generate SMT formula

### DISCUSSION THREAD

**Discussion by Sakura-0407** [binsec/binsec#46] [Q&A]
Hello, as far as I know, binsec uses SMT solvers for program validation. Does binsec support SMT formulas used by output validators?
For example: When I am validating my assembly code, can I generate the SMT formula corresponding to the validator?
```
.intel_syntax noprefix
    .globl Fr_copy
Fr_copy:

        mov     rax, [rsi + 0]
        mov     [rdi + 0], rax

        mov     rax, [rsi + 8]
        mov     [rdi + 8], rax

        mov     rax, [rsi + 16]
        mov     [rdi + 16], rax

        mov     rax, [rsi + 24]
        mov     [rdi + 24], rax

        mov     rax, [rsi + 32]
        mov     [rdi + 32], rax

        ret
    .type Fr_copy,@function
    .size Fr_copy,. - Fr_copy
```
```
rdi := nondet as dst            ## initialize the first argument
rsi := nondet as src            ## initialize the second argument

assume dst & 0b111 = 0          ## alignment constraint
assume src & 0b111 = 0          ## alignment constraint

assume 0x4000 <= dst <= 0x5000  ## arbitrary range  
assume 0x5000 <= src <= 0x6000  ## arbitrary range  

starting from <Fr_copy> with
  for i<64> in 0 to 4 do
    ## initialize the source content
    @[src + 8 * i, 8] := buf[8 * i, 8]
  end
end

replace <Fr_copy:last> by
  assert rdi = dst              ## check rdi final value
  assert rsi = src              ## check rsi final value
  eq<1> := true
  for i<64> in 0 to 4 do
    ## check the destination content
    eq := eq && @[dst + 8 * i, 8] = buf[8 * i, 8]
  end
  assert eq
  halt
end

explore all
```

**Comment by recoules** [comment]
Hi @MJJ-Shuai,

You can print the SMT formula of the path predicate at any point of the execution ([print formula](https://github.com/binsec/binsec/blob/master/doc/sse/references.md#print-full-smt-formula-print-formula)).

For instance, I would write something as the following.
```text
replace <Fr_copy:last> by
  ...
  assert eq
  print formula for eq
  halt
end
```
*In this example, the formula is trivially `true` and will be reduced to `1`.*

*Note that it will print as many formulas as there are paths in the program.*

Regards,
Frédéric

**Reply by Sakura-0407** [reply]
@recoules Thank you for such a quick reply, I tried to use print formula in my script, as you said, the formula is trivially true and will bereduced to 1.
But if i want to see the SMT formula corresponding to `@[dst +8*i,8] = buf[8*i,8]`, is there a way to make the output SMT more detailed?

**Reply by recoules** [reply]
Actually no, the syntactic rewriting is performed on the fly and memory accesses are simplified as soon as possible.

In your case, you may try to split the equality to show the content of memory accesses as follow. 

```text
replace <Fr_copy:last> by
  ...
  for i<64> in 0 to 4 do
    ## check the destination content
    print formula for @[dst + 8 * i, 8] as dst_content, buf[8 * i, 8] as src_content
    eq := eq && @[dst + 8 * i, 8] = buf[8 * i, 8]
  end
  assert eq
  halt
end
```

But the output is not really more informative, since ReadOverWrite have already took effect. Here, it is obvious that `dst_content` and `src_content` are equal (they are both defined with the same term `!5`).
```smt
(define-fun !1 () (_ BitVec 64) (_ bv20480 64))
(declare-fun src!2 () (_ BitVec 64))
(declare-fun dst!1 () (_ BitVec 64))
(define-fun !2 () (_ BitVec 64) (_ bv7 64))
(define-fun !3 () (_ BitVec 64) (_ bv0 64))
(define-fun !4 () (_ BitVec 64) (_ bv32 64))
(declare-fun buf () (Array (_ BitVec 64) (_ BitVec 8)))
(define-fun !5 () (_ BitVec 64) (concat (select buf (bvadd !4 (_ bv7 64))) (select buf (bvadd !4 (_ bv6 64))) (select buf (bvadd !4 (_ bv5 64))) (select buf (bvadd !4 (_ bv4 64))) (select buf (bvadd !4 (_ bv3 64))) (select buf (bvadd !4 (_ bv2 64))) (select buf (bvadd !4 (_ bv1 64))) (select buf !4)))
(define-fun dst_content () (_ BitVec 64) !5)
(define-fun src_content () (_ BitVec 64) !5)
(assert (and (bvule !1 src!2) (bvule src!2 (_ bv24576 64))))
(assert (and (bvule (_ bv16384 64) dst!1) (bvule dst!1 !1)))
(assert (= (bvand src!2 !2) !3))
(assert (= (bvand dst!1 !2) !3))
```

With more complex examples, however, you will likely obtain an SMT formula.
